cmd/chihaya/main.go

```diff
@@ -12,178 +11,138 @@ import (
 
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/spf13/cobra"
-	"gopkg.in/yaml.v2"
 
 	httpfrontend "github.com/chihaya/chihaya/frontend/http"
 	udpfrontend "github.com/chihaya/chihaya/frontend/udp"
 	"github.com/chihaya/chihaya/middleware"
 	"github.com/chihaya/chihaya/storage/memory"
 )
 
-type ConfigFile struct {
-	MainConfigBlock struct {
-		middleware.Config
-		PrometheusAddr string              `yaml:"prometheus_addr"`
-		HTTPConfig     httpfrontend.Config `yaml:"http"`
-		UDPConfig      udpfrontend.Config  `yaml:"udp"`
-		Storage        memory.Config       `yaml:"storage"`
-	} `yaml:"chihaya"`
-}
-
-// ParseConfigFile returns a new ConfigFile given the path to a YAML
-// configuration file.
-//
-// It supports relative and absolute paths and environment variables.
-func ParseConfigFile(path string) (*ConfigFile, error) {
-	if path == "" {
-		return nil, errors.New("no config path specified")
+func rootCmdRun(cmd *cobra.Command, args []string) error {
+	cpuProfilePath, _ := cmd.Flags().GetString("cpuprofile")
+	if cpuProfilePath != "" {
+		log.Println("enabled CPU profiling to " + cpuProfilePath)
+		f, err := os.Create(cpuProfilePath)
+		if err != nil {
+			return err
+		}
+		pprof.StartCPUProfile(f)
+		defer pprof.StopCPUProfile()
 	}
 
-	f, err := os.Open(os.ExpandEnv(path))
+	configFilePath, _ := cmd.Flags().GetString("config")
+	configFile, err := ParseConfigFile(configFilePath)
 	if err != nil {
-		return nil, err
+		return errors.New("failed to read config: " + err.Error())
```

Comment:
We could start using github.com/pkg/errors and make this to `errors.Wrap(err,"failed to read config")`. I use the package for pretty much every new project I start because I have to wrap so often.